src/ClientTransactionTask.cc

```diff
@@ -0,0 +1,562 @@
+                decision = WireFormat::TxDecision::ABORT;
+            }
+        } else if (rpc->responseHeader->status == STATUS_UNKNOWN_TABLET) {
+            // Nothing to do.  Will be retried.
+            TEST_LOG("STATUS_UNKNOWN_TABLET");
+        } else {
+            ClientException::throwException(HERE, rpc->responseHeader->status);
+        }
+
+        // Destroy object.
+        it = prepareRpcs.erase(it);
+    }
+}
+
+/**
+ * Send out a decision rpc if not all master have been notified.  Used in the
+ * commit method.  Factored out mostly for ease of testing.
+ */
+void
+ClientTransactionTask::sendDecisionRpc()
+{
+    // Issue an additional rpc.
+    DecisionRpc* nextRpc = NULL;
+    Transport::SessionRef rpcSession;
+    for (; nextCacheEntry != commitCache.end(); nextCacheEntry++) {
+        const CacheKey* key = &nextCacheEntry->first;
+        CacheEntry* entry = &nextCacheEntry->second;
+
+        if (entry->state == CacheEntry::DECIDE) {
+            continue;
+        }
+
+        if (nextRpc == NULL) {
+            rpcSession =
+                    ramcloud->objectFinder.lookup(key->tableId,
+                                                  key->keyHash);
+            decisionRpcs.emplace_back(ramcloud, rpcSession, this);
+            nextRpc = &decisionRpcs.back();
+        }
+
+        Transport::SessionRef session =
+                ramcloud->objectFinder.lookup(key->tableId, key->keyHash);
+        if (session->getServiceLocator() == rpcSession->getServiceLocator()
+            && nextRpc->reqHdr->participantCount <
+                    DecisionRpc::MAX_OBJECTS_PER_RPC) {
+            nextRpc->appendOp(nextCacheEntry);
+        } else {
+            break;
+        }
+    }
+    if (nextRpc) {
+        nextRpc->send();
+    }
+}
+
+/**
+ * Send out a prepare rpc if there are remaining un-prepared transaction ops.
+ * Used in the commit method.  Factored out mostly for ease of testing.
+ */
+void
+ClientTransactionTask::sendPrepareRpc()
+{
+    // Issue an additional rpc.
+    PrepareRpc* nextRpc = NULL;
+    Transport::SessionRef rpcSession;
+    for (; nextCacheEntry != commitCache.end(); nextCacheEntry++) {
+        const CacheKey* key = &nextCacheEntry->first;
+        CacheEntry* entry = &nextCacheEntry->second;
+
+        if (entry->state == CacheEntry::PREPARE) {
+            continue;
+        }
+
+        if (nextRpc == NULL) {
+            rpcSession =
+                    ramcloud->objectFinder.lookup(key->tableId,
+                                                  key->keyHash);
+            prepareRpcs.emplace_back(ramcloud, rpcSession, this);
+            nextRpc = &prepareRpcs.back();
+        }
+
+        Transport::SessionRef session =
+                ramcloud->objectFinder.lookup(key->tableId, key->keyHash);
+        if (session->getServiceLocator() == rpcSession->getServiceLocator()
+            && nextRpc->reqHdr->opCount < PrepareRpc::MAX_OBJECTS_PER_RPC) {
```

Comment:
Double-indent this line (and all continuation lines, unless you are using the "line up the arguments" approach).